### ISSUE TITLE
[keymgr, security] Assume adversarial SW model regarding keymgr analysis

### ISSUE THREAD

**Issue by ballifatih** [lowRISC/opentitan#17585]
### Description

This is a security opinion on threat modeling regarding keymgr, which spun off from an offline discussion with @gdessouky and @bilgiday and is closely related to #17508. I wanted to have a separate issue to discuss this, so that I can easily refer to this from other dependent issues.

TL;DR: I think when we are analyzing security issues involving `keymgr`, we should assume that all layers of SW are controlled by an attacker (including ROM and ROMExt).

`keymgr` is responsible for keeping device-unique secrets secure. These secrets include those that are received from other IPs (OTP root key, flash diversification keys etc.) as well as those generated internally through KMAC calls (internal state, attestation seeds, sealing encryption keys, OTBN seeds etc.) My interpretation is the raison d'être for `keymgr` is that the secrecy of these sensitive key material is guaranteed by HWIP itself without relying on any assumption from SW side. Even if SW is compromised, `keymgr` should deliver on the promise that these secrets will not be compromised. Maybe these keys were used to encrypt sensitive user data locally and we would like to keep it secure even if the platform is compromised.

This might seem to contradict with the fact that HWIPs + ROM + ROMExt are all owned by a SiliconCreator, so from that point of view, why would SiliconCreator be considered as an attacker? The security OpenTitan can deliver if SiliconCreator is adversarial is limited and out of our scope, and SiliconCreator could come up with many other ways to reach its malicious goal without having to touch any layer of SW.

Here, I can imagine few things possibly going wrong even if SiliconCreator does everything right:

1. There might be some bugs in ROM, ROMExt or upper layer of SW that might allow attacker to partially or fully control Ibex. For instance, maybe attacker has a way of weakening randomness used by `keymgr`'s internal LFSR and combine with SCA #15088.
2. Even if there are no bugs in ROM, ROMExt, the drivers already need to tackle lots of issues of their own. If HW does not prevent particular potential weakness, patching it through drivers requires a tedious analysis and even then we might miss few attack scenarios.
3. In the worst case scenario, if secure boot is compromised (e.g. SiliconOwner root key stolen, and that key is not revoked yet), I think this is not a reason for `keymgr` to not deliver on its promise of protecting its keys. 

In short, the main point is that the security of `keymgr` should not rely on so many assumptions outside HWIP itself.

Let me know if this threat modeling sounds too far-fetched before I dive deeper into analyzing `keymgr` and `kmac` interaction in this model. Of course this is orthogonal to M2.5 bucketing, and many potential issues arising from that point of view could be tagged as FutureRelease.

cc: @johannheyszl @moidx @vogelpi @jadephilipoom

**Comment by johngt** [comment]
Also cc'ing @alphan / @luismarques based on WG meeting

**Comment by gdessouky** [comment]
Agree, as we discussed, except for ROM (also for key manager). The trust has to start from some assumptions, besides the HWIPs, as minimal as possible. I think it is safe and justified (when very heavily verified and tested) to include ROM in the TCB, even in the most capable SW attacker scenario. 

I know your discussion here focuses on key manager, which is also another point. It's hard to consistently assume a very capable threat model for all assets and scenarios of the chip; there are some issues where you end up counting on ROM_EXT still. It will make a potential attack more difficult, but still a risk, which can also be justified by verifying the bit that you rely on for something security-critical very extensively. So I am starting to view that the trust assumptions/threat model would also have to depend on the assets/security impact in question, and not consistent for all design decisions in your chip - not sure if that's the right approach though :)

**Comment by cdgori** [comment]
Thank you for writing this up.

> My interpretation is the raison d'être for keymgr is that the secrecy of these sensitive key material is guaranteed by HWIP itself without relying on any assumption from SW side. Even if SW is compromised, keymgr should deliver on the promise that these secrets will not be compromised. Maybe these keys were used to encrypt sensitive user data locally and we would like to keep it secure even if the platform is compromised.

This is my preferred way to consider the functionality of keymgr as well, and was what Tim and I talked about in the earliest stages of the architecture/design of the block.

I also prefer this adversarial SW model (in general, but in particular for keymgr), to try to protect against failures in the software stack.  However, I would accept that perhaps we have to consider the ROM as ultimately trusted.  (It would be better if we did not, but it may not be feasible, and we will have to do the thorough testing described by @gdessouky in order to include ROM in the TCB.)

**Comment by johannheyszl** [comment]
This issue describes a threat model to use for future keymgr analysis. I am thus labeling accordingly
